### PR TITLE
ARROW-5111: [Go] implement reading list arrays from Arrow file

### PR DESCRIPTION
**Pull request by sbinet**

### CODE REVIEW

**Comment by sbinet** [comment]
needs #4155

File: go/arrow/ipc/file_reader.go [hunk 1]
```diff
@@ -409,6 +425,29 @@ func (ctx *arrayLoaderContext) loadPrimitive(dt arrow.DataType) array.Interface
 	return array.MakeFromData(data)
 }
 
+func (ctx *arrayLoaderContext) loadBinary(dt arrow.DataType) array.Interface {
+	field, buffers := ctx.loadCommon(3)
```
Comment (nevi-me) [inline go/arrow/ipc/file_reader.go]: Hi @sbinet, I'm working on the equivalent for Rust. Hope you don't mind me asking a few questions.

* Does the binary array always consist of 3 buffers, or would it be 2 if the null count is 0? Here's the relevant code in the Rust impl: https://github.com/apache/arrow/pull/4167/files#diff-cd3519a5e748548b5323b0d05d8e9c8aR53 (cc @pitrou @kszucs )
* Do you mind sharing a script for generating Arrow data in Python (I took the one you used when you asked a question in the mailing list)? I haven't used pyarrow in a while, so I was struggling with making lists and structs.

Thanks

Comment (pitrou) [reply go/arrow/ipc/file_reader.go]: A binary array always has 3 buffers, but the first buffer may be null or empty if the null count is 0.

This means buffer #1 should always be the offsets array and buffer #2 the values array.

Comment (nevi-me) [reply go/arrow/ipc/file_reader.go]: Thanks @pitrou, is it safe for me to presume that the buffer numbers for primitives, list and struct will behave in the same way, where the null buffer may be null or empty?

Comment (pitrou) [reply go/arrow/ipc/file_reader.go]: Yes, it's part of the Arrow format specification.

Comment (sbinet) [reply go/arrow/ipc/file_reader.go]: @nevi-me here is what I am using:
https://gist.github.com/sbinet/8ec09eb95bc89c6d71cc192dfcce677b

**Comment by sbinet** [comment]
PTAL @stuartcarnie @alexandreyc @nickpoorman

**Review by stuartcarnie — APPROVED** [review]

**Pull request closed without merging**